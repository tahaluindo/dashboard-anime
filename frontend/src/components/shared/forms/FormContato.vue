<template>
  <form ref="form">
    <form-field
      labelName="Nome"
      :required="true"
      v-model:data-from-parent="name"
      :lazy="true"
    />
    <form-field labelName="Email" v-model:data-from-parent="email" />
    <form-field labelName="Telefone" v-model:data-from-parent="telephone" />
    <form-text-area
      labelName="Descrição"
      v-model:data-from-parent="description"
      height="150px"
    />
    <the-button
      class="
        btn-xl-border-md
        font-opensans font-bold
        clamptext-1-14
        py-1
        justify-self-center
        hover:text-white
        hover:bg-primary
      "
      @click.prevent="sendMessage"
    >
      <p>ENVIAR</p>
    </the-button>
  </form>
</template>

<script lang="ts">
import { defineComponent, ref } from "vue";
import FormField from "./FormField.vue";
import FormTextArea from "./FormTextArea.vue";
import FormDefault from "./FormDefault.vue";
import TheButton from "../button/TheButton.vue";

export default defineComponent({
  components: { FormField, FormTextArea, FormDefault, TheButton },
  setup() {
    const name = ref("");
    const email = ref("");
    const telephone = ref("");
    const description = ref("");
    const form = ref(null as unknown);

    const sendMessage = async () => {
      const formData = new FormData(
        form.value as HTMLFormElement
      ) as URLSearchParams;

      try {
        await fetch("/", {
          method: "POST",
          headers: { "Content-Type": "application/x-www-form-urlencoded" },
          body: new URLSearchParams(formData).toString(),
        });

        window.alert("Formulário enviado com sucesso!");
      } catch (error) {
        window.alert("Erro ao enviar formulário!");
      }
    };

    return {
      name,
      email,
      telephone,
      description,
      sendMessage,
      form,
    };
  },
});
</script>
