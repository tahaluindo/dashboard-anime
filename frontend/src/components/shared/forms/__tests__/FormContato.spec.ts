import { flushPromises, mount } from "@vue/test-utils";
import {
  mockFetch,
  mockFetchRejected,
} from "../../../../__mocks__/global.mock";
import { mockWindowAlert } from "../../../../__mocks__/window.mock";
import FormContato from "../FormContato.vue";

describe("FormContato component test", () => {
  it("should render the default component.", () => {
    const wrapper = mount(FormContato);

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should send the form when click the button submit.", async () => {
    mockWindowAlert();
    mockFetch();

    const wrapper = mount(FormContato);

    const windowAlertSpy = jest.spyOn(window, "alert");

    await wrapper.find("button").trigger("click.prevent");
    await flushPromises();

    expect(windowAlertSpy).toHaveBeenCalledWith(
      "Formulário enviado com sucesso!"
    );
  });

  it("should error if the fetch fails.", async () => {
    mockWindowAlert();
    mockFetchRejected();

    const wrapper = mount(FormContato);

    const windowAlertSpy = jest.spyOn(window, "alert");

    await wrapper.find("button").trigger("click.prevent");
    await flushPromises();

    expect(windowAlertSpy).toHaveBeenCalledWith("Erro ao enviar formulário!");
  });
});
