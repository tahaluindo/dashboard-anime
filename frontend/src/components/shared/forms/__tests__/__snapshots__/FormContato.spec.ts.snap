// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormContato component test should render the default component. 1`] = `
"<form>
  <div class=\\"relative h-16 w-full\\"><input class=\\"bg-none p-3 pb-1 w-full border-b-2 border-t-0 border-r-0 border-l-2 border-current text-current rounded focus:border-secondary focus:ring-0\\" type=\\"text\\" name=\\"Nome\\" id=\\"Nome\\"><label class=\\"absolute rounded top-1 on-focus left-3 text-current bg-none z-10 p-1 duration-450 decorator-required\\" for=\\"Nome\\">Nome</label></div>
  <div class=\\"relative h-16 w-full\\"><input class=\\"bg-none p-3 pb-1 w-full border-b-2 border-t-0 border-r-0 border-l-2 border- border-current text-current rounded focus:border-secondary focus:ring-0\\" type=\\"text\\" name=\\"Email\\" id=\\"Email\\"><label class=\\"absolute rounded top-1 on-focus left-3 text-current bg-none z-10 p-1 duration-450\\" for=\\"Email\\">Email</label></div>
  <div class=\\"relative h-16 w-full\\"><input class=\\"bg-none p-3 pb-1 w-full border-b-2 border-t-0 border-r-0 border-l-2 border- border-current text-current rounded focus:border-secondary focus:ring-0\\" type=\\"text\\" name=\\"Telefone\\" id=\\"Telefone\\"><label class=\\"absolute rounded top-1 on-focus left-3 text-current bg-none z-10 p-1 duration-450\\" for=\\"Telefone\\">Telefone</label></div>
  <div class=\\"relative w-full\\"><textarea class=\\"bg-none p-3 pb-1 w-full border-b-2 border-t-0 border-r-0 border-l-2 border- border-current text-current rounded focus:border-secondary focus:ring-0\\" style=\\"height: 150px;\\" type=\\"text\\" name=\\"Descrição\\" id=\\"Descrição\\"></textarea><label class=\\"absolute rounded top-1 on-focus left-3 text-current bg-none z-10 p-1 duration-450\\" for=\\"Descrição\\">Descrição</label></div><button class=\\"btn-xl-border-md
        font-opensans font-bold
        clamptext-1-14
        py-1
        justify-self-center
        hover:text-white
        hover:bg-primary\\">
    <p>ENVIAR</p>
  </button>
</form>"
`;
